src/Type/SubtractedType.php

```diff
@@ -0,0 +1,313 @@
+			return $originalType;
+		}
+
+		return new self($originalType, $subtractedType);
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function getReferencedClasses(): array
+	{
+		return array_merge(
+			$this->originalType->getReferencedClasses(),
+			$this->subtractedType->getReferencedClasses()
+		);
+	}
+
+	public function accepts(Type $type, bool $strictTypes): TrinaryLogic
+	{
+		$original = $this->originalType->accepts($type, $strictTypes);
+		if (!$original->yes()) {
+			return $original;
+		}
+
+		return TrinaryLogic::createFromBoolean(!$this->subtractedType->equals($type));
+	}
+
+	public function isSuperTypeOf(Type $type): TrinaryLogic
+	{
+		// TODO: Implement isSuperTypeOf() method.
+	}
+
+	public function equals(Type $type): bool
+	{
+		if (!$type instanceof self) {
+			return false;
+		}
+
+		return $this->originalType->equals($type) && !$this->subtractedType->equals($type);
+	}
+
+	public function describe(VerbosityLevel $level): string
+	{
+		return $this->originalType->describe($level) . '~' . $this->subtractedType->describe($level);
+	}
+
+	public function canAccessProperties(): TrinaryLogic
+	{
+		if ($this->subtractedType->equals(new ObjectWithoutClassType())) {
+			return TrinaryLogic::createNo();
+		}
+
+		if ($this->originalType instanceof MixedType) {
+			return TrinaryLogic::createYes();
+		}
+
+		$original = $this->originalType->canAccessProperties();
+		if (!$original->yes()) {
+			return $original;
+		}
+
+		if ($this->subtractedType->canAccessProperties()->no()) {
+			return TrinaryLogic::createYes();
+		}
+
+		/*
+		 * TODO:
+		 * Both the original and the subtracted type can access properties
+		 * now we need to determine if *only* the subtracted type has it, or if it is
+		 * present without the subtracted type as well.
+		 */
+		return TrinaryLogic::createMaybe();
+	}
+
+	public function hasProperty(string $propertyName): TrinaryLogic
+	{
+		if ($this->subtractedType->equals(new ObjectWithoutClassType())) {
+			return TrinaryLogic::createNo();
+		}
+
+		if ($this->originalType instanceof MixedType) {
+			return TrinaryLogic::createYes();
+		}
+
+		$original = $this->originalType->hasProperty($propertyName);
+		if (!$original->yes()) {
+			return $original;
+		}
+
+		if ($this->subtractedType->hasProperty($propertyName)->no()) {
+			return TrinaryLogic::createYes();
+		}
+
+		/*
+		 * TODO:
+		 * Both the original and the subtracted type have the property,
+		 * now we need to determine if *only* the subtracted type has it, or if it is
+		 * present without the subtracted type as well.
+		 */
+		return TrinaryLogic::createMaybe();
+	}
+
+	public function getProperty(string $propertyName, ClassMemberAccessAnswerer $scope): PropertyReflection
+	{
+		/*
+		 * TODO: possible validate this
+		 * Assume for now that this is only called if the property is present, thus we can retrieve
+		 * it from the original type.
+		 */
+		return $this->originalType->getProperty($propertyName, $scope);
```

Comment:
Is it okay to just retrieve the original property? Prehaps we could check here again, and throw a shouldNotHappenException here.